rclcpp_lifecycle/src/lifecycle_node_interface_impl.hpp

```diff
@@ -411,27 +451,32 @@ class LifecycleNode::LifecycleNodeInterfaceImpl
         return rcl_lifecycle_transition_error_label;
       };
 
-    cb_return_code = execute_callback(state_machine_.current_state->id, initial_state);
+    cb_return_code = execute_callback(current_state_id, initial_state);
     auto transition_label = get_label_for_return_code(cb_return_code);
 
-    if (
-      rcl_lifecycle_trigger_transition_by_label(
-        &state_machine_, transition_label, publish_update) != RCL_RET_OK)
     {
-      RCUTILS_LOG_ERROR(
-        "Failed to finish transition %u. Current state is now: %s (%s)",
-        transition_id, state_machine_.current_state->label, rcl_get_error_string().str);
-      rcutils_reset_error();
-      return RCL_RET_ERROR;
+      std::lock_guard<std::recursive_mutex> lock(state_machine_mutex_);
+      if (
+        rcl_lifecycle_trigger_transition_by_label(
+          &state_machine_, transition_label, publish_update) != RCL_RET_OK)
+      {
+        RCUTILS_LOG_ERROR(
+          "Failed to finish transition %u. Current state is now: %s (%s)",
+          transition_id, state_machine_.current_state->label, rcl_get_error_string().str);
+        rcutils_reset_error();
+        return RCL_RET_ERROR;
+      }
+      current_state_id = state_machine_.current_state->id;
     }
 
     // error handling ?!
     // TODO(karsten1987): iterate over possible ret value
     if (cb_return_code == node_interfaces::LifecycleNodeInterface::CallbackReturn::ERROR) {
       RCUTILS_LOG_WARN("Error occurred while doing error handling.");
 
-      auto error_cb_code = execute_callback(state_machine_.current_state->id, initial_state);
+      auto error_cb_code = execute_callback(current_state_id, initial_state);
       auto error_cb_label = get_label_for_return_code(error_cb_code);
+      std::lock_guard<std::recursive_mutex> lock(state_machine_mutex_);
```

Comment:
I think it would be better to keep holding the lock across the above and down into this code.  This minimizes the number of times we take and drop the lock in this function.

Reply:
@fujitatomoya 

About this problem, I discussed with iuhilnehc-ynos last Friday.  If we want to resolve this completely, the current idea cannot avoid introducing thread and queue. But this make code more complex (may introduce new problem). Another way is to keep this solution, and we need to remind users to avoid changing the node state in different threads.

Reply:
> for example,
> 
> 1. thread-1 calls `trigger_transition` and hold `state_machine_mutex_` internally.
> 2. context switch to thread-2 holds user-mutex, and call `trigger_transition`.
> 3. thread-2 is waiting on `state_machine_mutex_`.
> 4. context switch to thread-1, and user callback requires user-mutex.
> 5. deadlock

Sorry, I noticed that I made a mistake in https://github.com/ros2/rclcpp/pull/1756#discussion_r738962746 which still leads to a deadlock by following this example.

Back to this example, I think it seems to be the user's responsibility not to do these actions.
There is no reason to hold user-mutex to call `trigger_transition`  on step 2 if the user wants to hold the user-mutex inside user callback.

I am inclined to keep the original lock. Let's hear others' ideas.

Reply:
yeah, basically i agree with the idea you mentioned. but i think error state is much better than deadlock. and we cannot know how user application is developed. with all these possible conditions and we already have the similar code in action, i still think we need to release the lock during user callback.

https://github.com/ros2/rclcpp/blob/152dbc6e38e18c18a5883b51c33162d237f891e0/rclcpp_action/src/server.cpp#L411

Reply:
The more I look at `get_current_state`, the less sense it makes to me.  It really should be a `const` method, but it can't be because it modifies internal state of the class.  I think we should address that problem first before we go forward here.

Towards that end, I've opened up https://github.com/ros2/rclcpp/pull/2027 .  This doesn't actually fix the problem, but does a number of cleanups in this class and should make it easier for us to make that method const in a followup.

Reply:
@clalancette thanks, i will look at another PR, then we can get back to this if necessary.